### ISSUE TITLE
`grid_layout` not available anymore

### ISSUE THREAD

**Issue by giovp** [napari/napari#4709]
while trying to access `grid_layout` with

`viewer.window.qt_viewer.controls.widgets[viewer.layers["CD36:X:image1"]].grid_layout`

I get:
```
/var/folders/3t/lc9m5zv966934dq4d6gf6l60g_dj6h/T/ipykernel_36457/1739642263.py:1: FutureWarning: Public access to Window.qt_viewer is deprecated and will be removed in
v0.5.0. It is considered an "implementation detail" of the napari
application, not part of the napari viewer model. If your use case
requires access to qt_viewer, please open an issue to discuss.
  viewer.window.qt_viewer.controls.widgets[viewer.layers["CD36:X:image1"]].grid_layout
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [37], in <cell line: 1>()
----> 1 viewer.window.qt_viewer.controls.widgets[viewer.layers["CD36:X:image1"]].grid_layout

AttributeError: 'QtPointsControls' object has no attribute 'grid_layout'
```

yet it seems it's still in the docs?

https://github.com/napari/napari/blob/e0cc713d4454524093bafdd3507ab18c9f86fc2a/napari/_qt/layer_controls/qt_points_controls.py#L40

on another note: I've used to access `grid_layout` to be able to add widgets to the layer controls, what is the recommended way to do it now? 

this is an example taken from here: https://squidpy.readthedocs.io/en/stable/external_tutorials/tutorial_napari.html

![image](https://user-images.githubusercontent.com/25887487/174435217-980db2c0-c1c5-42d9-904b-4973cb1a5682.png)

here, the "percentile" widget added to the points layer controls the range of the colormap visualized at the bottom left of the napari GUI. 

In short I'd like to add widgets to the layers controls of newly created layer by the plugin.

thanks in advance for the help!

**Comment by Czaki** [comment]
> here, the "percentile" widget added to the points layer controls the range of the colormap visualized at the bottom left of the napari GUI.

viewer.window.add_dock_widget

https://github.com/napari/napari/blob/e0cc713d4454524093bafdd3507ab18c9f86fc2a/napari/_qt/qt_main_window.py#L729-L775

**Comment by tlambert03** [comment]
Yeah, I’m afraid modifying the layer controls in a plugin isn’t supported at the moment. As the warning you’re seeing indicates _qt_viewer is private and not intended for access (esp by plugins).  This use case would require a fundamentally different/new access point for plugins 

(you’re free of course to do it as an end user of napari, but not currently as a plugin modifying other peoples viewers)

**Comment by giovp** [comment]
thanks for the prompt reply!

> > here, the "percentile" widget added to the points layer controls the range of the colormap visualized at the bottom left of the napari GUI.
> 
> viewer.window.add_dock_widget
> 
> https://github.com/napari/napari/blob/e0cc713d4454524093bafdd3507ab18c9f86fc2a/napari/_qt/qt_main_window.py#L729-L775

this is what I'm doing atm but it gets messy really fast (as layer are added, more sliders and cmaps are added as well) and it gets tricky to track layer states.
![image](https://user-images.githubusercontent.com/25887487/174452457-a0231f34-292b-4123-90c6-f0ce0b656d8a.png)

> Yeah, I’m afraid modifying the layer controls in a plugin isn’t supported at the moment. As the warning you’re seeing indicates _qt_viewer is private and not intended for access (esp by plugins). This use case would require a fundamentally different/new access point for plugins

I get why is private and shouldn't be modified by plugins, but it'd be nice to have a way to "dock" widgets states to layers.

**Comment by Czaki** [comment]
> this is what I'm doing atm but it gets messy really fast (as layer are added, more sliders and cmaps are added as well) and it gets tricky to track layer states.

You may want to connect `viewer.layers.selection.events.active` event and update widget content base on current active layer.

**Comment by tlambert03** [comment]
> but it'd be nice to have a way to "dock" widgets states to layers.

elaborate?  By dock do you mean lock or otherwise link such that the widget updates itself when the layer does?

**Comment by brisvag** [comment]
If I understand correctly, the goal is to have a *single* widget, but have it always refer to the selected layer. Then yes, I would do as @Czaki suggests, as there is no "official" way to do this at the moment.

This use case was actually mentioned as one of the reasons in favor of the tab layout for layer controls in #4381, as it would make the interface for this kind of thing a lot easier.

**Comment by brisvag** [comment]
Btw, a random thought: It looks like the thing you're adding is a colormap indicator of some sort. It would be *awesome* if we had a way to add things like that directly on the canvas, just like the scale bar...

**Comment by Czaki** [comment]
> Btw, a random thought: It looks like the thing you're adding is a colormap indicator of some sort. It would be _awesome_ if we had a way to add things like that directly on the canvas, just like the scale bar...

it looks like support for #3691.

**Comment by brisvag** [comment]
Not sure if that's a good candidate for technical layers. This is an overlay (i.e: it should not live in *world* space, but *canvas* space).

**Comment by Czaki** [comment]
True

**Comment by giovp** [comment]
thanks all for the discussion! Let me answer briefly:

> elaborate? By dock do you mean lock or otherwise link such that the widget updates itself when the layer does?

maybe wrong word but by dock I mean to actually add the widget to the layers control panel (as was available before 
see image in here: https://github.com/napari/napari/issues/4709#issue-1275764819 where "percentile" would be the slider controlling the clipping and the colormap). This way, by changing the layer, the percentile slider would visualize in the right state when it was set for that specific layer.

> You may want to connect viewer.layers.selection.events.active event and update widget content base on current active layer.

this is indeed an option that should be working. As of now the min-max and clipping is already saved in the layers metadata so I could just updated as the selected layer changes.

> If I understand correctly, the goal is to have a single widget, but have it always refer to the selected layer. Then yes, I would do as @Czaki suggests, as there is no "official" way to do this at the moment.

yes or also different instances of the same widget per layer.

> This use case was actually mentioned as one of the reasons in favor of the tab layout for layer controls in https://github.com/napari/napari/pull/4381, as it would make the interface for this kind of thing a lot easier.

this looks very interesting!

> Btw, a random thought: It looks like the thing you're adding is a colormap indicator of some sort. It would be awesome if we had a way to add things like that directly on the canvas, just like the scale bar...

If you think it's interesting I'd be happy to contribute but I'm not sure it's actually useful without the slider widget. With the slider you can clip the real values and hence the colormap values, which also modify the points/labels in real time. I think the scalebar instead respond to the zooming of the canvas? Never used it but I should!

another option maybe useful would be to have the colormap + min-max values of the features plotted + the slider in the layers control directly (and not the canvas like scalebar). Not sure if this would be general enough to be a separate widget in the layer control.

**Comment by tlambert03** [comment]
>  by dock I mean to actually add the widget to the layers control panel (as was available before)

ah, ok.  Well yeah, that's a thing (access to the private _qt_viewer model) that's specifically not allowed in plugins anymore.

however... that doesn't mean that you can't still achieve what I think you mean by "This way, by changing the layer, the percentile slider would visualize in the right state when it was set for that specific layer.".  You just need to connect the appropriate [layer events](https://napari.org/guides/events_reference.html) to a method that updates your slider widget.  In other words, there's nothing "special" about a widget that gets put in the layer controls.  You can achieve the same sort of updating anywhere.

**Comment by psobolewskiPhD** [comment]
> > Btw, a random thought: It looks like the thing you're adding is a colormap indicator of some sort. It would be awesome if we had a way to add things like that directly on the canvas, just like the scale bar...
> 
> If you think it's interesting I'd be happy to contribute but I'm not sure it's actually useful without the slider widget. With the slider you can clip the real values and hence the colormap values, which also modify the points/labels in real time. I think the scalebar instead respond to the zooming of the canvas? Never used it but I should!

I think the idea is to have a color bar, that shows the colormap lookup table values.
So it would reflect what the numerical values of the underlying data are for the color shades using the colormap and contrast limits.
I also think this is valuable—in ImageJ this is called the `Calibration Bar`: 
https://imagej.nih.gov/ij/docs/menus/analyze.html#cbar
It could also be used to provide depth coding information for modalities like AFM.
Separate PR from this issue though for sure!
